src/substituter.ts

```
@@ -0,0 +1,106 @@
+          operator,
+          left,
+          reducedRight as es.Expression,
+          node.loc!
+        )
+        return [reducedExpression, context]
+      }
+    } else {
+      const [reducedLeft] = reduce(node.left, context)
+      const reducedExpression = ast.binaryExpression(
+        operator,
+        reducedLeft as es.Expression,
+        right,
+        node.loc!
+      )
+      return [reducedExpression, context]
+    }
+  },
+  Program(node: es.Program, context: Context) {
+    const [firstStatement, ...otherStatements] = node.body
+    if (
+      firstStatement.type === 'ExpressionStatement' &&
+      firstStatement.expression.type === 'Literal'
+    ) {
+      return [ast.program(otherStatements, node.loc!), context]
+    } else {
+      const [reduced, newContext] = reduce(firstStatement, context)
+      return [ast.program([reduced as es.Statement, ...otherStatements], node.loc!), newContext]
+    }
+  },
+  /* <BAD PRACTICE>
+   * for now BlockStatement is just duplicated code from Program which is BAD PRACTICE,
+   * but in future BlockStatement will need to handle return as well.
+   * </BAD PRACTICE>
+   */
+  BlockStatement(node: es.BlockStatement, context: Context) {
+    const [firstStatement, ...otherStatements] = node.body
+    const isFirstStatementCompletelyReduced =
+      firstStatement.type === 'ExpressionStatement' && firstStatement.expression.type === 'Literal'
+    if (isFirstStatementCompletelyReduced) {
+      return [ast.blockStatement(otherStatements, node.loc!), context]
+    } else {
+      const [reduced, newContext] = reduce(firstStatement, context)
+      return [
+        ast.blockStatement([reduced as es.Statement, ...otherStatements], node.loc!),
+        newContext
+      ]
+    }
+  }
+}
+
+function reduce(node: es.Node, context: Context): [es.Node, Context] {
+  const reducer = reducers[node.type]
+  if (reducer === undefined) {
+    return [node, context] // if reducer is not found we just get stuck
```

I think this should throw an error or otherwise signal the end of the reduction.

---

Yeah. I'll throw an irreducible exception